### PR TITLE
Updates to onchip_memory_cache tutorial

### PR DESCRIPTION
**Pull request by mdbtucker**
## Description

Preliminary review for rewrite of the onchip_memory_cache tutorial.

## Type of change

Please delete options that are not relevant. Add a 'X' to the one that is applicable. 

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Implement fixes for ONSAM Jiras

## How Has This Been Tested?
 - running regtests to confirm performance at various cache depths 
 - cache depth 0 (II2 with load latency 1): https://spetc.intel.com/testsummary?testRunIds=6561034
 - cache depth 1 (II1 with load latency 1): https://spetc.intel.com/testsummary?testRunIds=6561057
 - cache depth 7 (II1 with load latency 4): https://spetc.intel.com/testsummary?testRunIds=6561065
 - testing result: FMAX shows no change or slight improvement of this technique over ivdep (S10 with cache ~5% better FMAX, S10 without cache and A10 with and without cache basically unchanged)
 - ran Linux regtest with default parameters, all tests passed
 - running Windows regtests on emulation, a10_ref and s10_ref, waiting for results (had some issues launching the test)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # CMAKE build directories
-build/
+build*/
```
Comment (yuguen) [inline .gitignore]: This is definitely helpful, but:
1/ This is outside the FPGA part of the repo
2/ Not really related to the onchip_memory_cache tutorial
Should this be in a different PR?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/CMakeLists.txt [hunk 2]
```diff
@@ -19,16 +19,32 @@ endif()
...
 #    For this reason, FPGA backend flags must be passed as link flags in CMake.
-set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fintelfpga -DFPGA_EMULATOR")
-set(EMULATOR_LINK_FLAGS "-fintelfpga")
-set(HARDWARE_COMPILE_FLAGS "-Wall ${WIN_FLAG} -fintelfpga")
-set(HARDWARE_LINK_FLAGS "-fintelfpga -Xshardware -Xsboard=${FPGA_BOARD} ${USER_HARDWARE_FLAGS}")
+set(EMULATOR_COMPILE_FLAGS "-Wall ${WIN_FLAG} ${CACHE_DEPTH_FLAG} ${AC_TYPES_FLAG} -fintelfpga -DFPGA_EMULATOR")
+set(EMULATOR_LINK_FLAGS "${AC_TYPES_FLAG} -fintelfpga")
+set(HARDWARE_COMPILE_FLAGS "-Wall ${WIN_FLAG} ${CACHE_DEPTH_FLAG} ${AC_TYPES_FLAG} -fintelfpga")
+set(REPORT_LINK_FLAGS "-fintelfpga -Xshardware -Xsboard=${FPGA_BOARD} ${CACHE_DEPTH_FLAG} ${USER_HARDWARE_FLAGS}")
+set(HARDWARE_LINK_FLAGS "${REPORT_LINK_FLAGS} ${AC_TYPES_FLAG}")
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/CMakeLists.txt:47]: I think you should use the AC_TYPES_FLAG for the report as well. This way you would only have REPORT_LINK_FLAGS and not HARDWARE_LINK_FLAGS, which would be more similar to the other designs cmake files

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 3]
```diff
@@ -5,140 +5,97 @@
...
 #include <chrono>
 
+#include "cached_local_memory.hpp"  // DirectProgramming/DPC++FPGA/include
+#include "unrolled_loop.hpp"
+
 // dpc_common.hpp can be found in the dev-utilities include folder.
 // e.g., $ONEAPI_ROOT/dev-utilities//include/dpc_common.hpp
 #include "dpc_common.hpp"
 
-using namespace sycl;
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:14]: Thanks for making this change as part of this work!

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 4]
```diff
@@ -5,140 +5,97 @@
...
       }
 
       // Write output to global memory
       for (uint32_t b = 0; b < kNumOutputs; ++b) {
-        if (!use_cache) {
-          output[b] = local_output[b];
-        } else {
-          output[b] = local_output_with_cache[b];
-        }
+        output[b] = histogram.read(b);
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp]: You did a similar change below: you can rename `b` to `hist_group`

File: DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+ public:
+  static constexpr int kNumAddrBits = fpga_tools::CeilLog2(k_mem_depth);
+  using addr_t = ac_int<kNumAddrBits, false>;
+
+  CachedLocalMemory() {
+    for (int i = 0; i < k_cache_depth; i++) {
+      cache_valid_[i] = false;
+    }
+  }
+  CachedLocalMemory(T init_val) {
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp]: What is the point of having an initializer if all the valids are false?

File: DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp [hunk 6]
```diff
@@ -87,7 +87,7 @@ constexpr void UnrolledLoop(F&& f) {
 }
 
 //
-// Convenience implementation (C)
+// Convenience implementation (D)
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp:90]: What this intended?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 7]
```diff
@@ -157,70 +114,70 @@ int main() {
...
-        default: {
-          Histogram<false>(q, input_buf, output_buf, e);
+    // iterate over the cache depths
+    for (int i = kMinCacheDepth; i < kMaxCacheDepth + 1; i++) {
+
+      std::cout << "Beginning run with cache depth " << i;
+      if (i == 0) { std::cout << " (no cache)"; }
+      std::cout << std::endl;
+
+      fpga_tools::UnrolledLoop<kMinCacheDepth, kMaxCacheDepth+1>([&](auto j) {
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:135]: Interesting workaround.
This is not an obvious technique that we expect the user to understand (and the underlying reasons) just by having a look at the code of this tutorial. 
I think you should add a comment explaining why you did that, and that it has nothing to do with how you should use the histogram

**Comment by yuguen** [comment]
You probably though about it already, but there are _two_ readmes to update: the tutorial readme and the include readme :)

**Comment by yuguen** [comment]
You also probably need to update the MSVS files to have the added include path to the include folder

File: .gitignore (hunk 1 above)
Comment (mdbtucker) [reply .gitignore]: Agreed.  Reverted this change.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 3 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:14]: You're welcome :-)

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 4 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp]: Replaced all uses of b with hist_group

File: DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp (hunk 5 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp]: Setting the valids to false just means you ignore the cache.  The initializer sets all the values in the actual memory.  So for the histogram example, I initialize the memory to 0, which is required for correct operation.

File: DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp (hunk 6 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp:90]: Yes it was.  It's not really related to this change, just something I noticed while working on this change, a simple error in the comments of this file.  I could put it into a separate PR if you want, but it's small enough I was hoping to just sneak it through with this one . . .

File: DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp (hunk 5 above)
Comment (yuguen) [reply DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp]: Why it is required? 
If the valids are 0, then no matter what, the data at that location is never going to be read
So it does not matter if it is initialized or not, does it?

File: DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp (hunk 6 above)
Comment (yuguen) [reply DirectProgramming/DPC++FPGA/include/unrolled_loop.hpp:90]: I'm fine with the sneaking :)

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 7 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:135]: Good point, I added a comment.  Have a look and please resolve this comment if you think it's sufficient.

**Comment by mdbtucker** [comment]
> You probably though about it already, but there are _two_ readmes to update: the tutorial readme and the include readme :)

Of course I thought of that!  Why I'm offended at the very implication that I might have forgotten!  How dare you sir?  HOW DARE YOU?!?

(ok, maybe I forgot a tiny bit)

**Comment by mdbtucker** [comment]
I spoke with Joe about using the term 'local memory', and he pointed out it's overly restrictive (this technique can apply to private memory, and to device globals).  We agreed that the term 'onchip memory' is more accurate.

I still feel that 'onchip_memory_cache' is a bit ambiguous - it could mean just the cache is in onchip memory and the memory itself is off chip, or it could mean the memory BEING cached is on chip.  So I've renamed my class to 'OnchipMemoryWithCache', and renamed the header file to match.

I feel like this is close enough to 'onchip_memory_cache' that I don't have to rename the tutorial.  I think renaming the tutorial is a bit of a pain and has some trickle-down extra work involved that I'd rather avoid.  Let me know if you don't like this compromise.

**Comment by yuguen** [comment]
> I spoke with Joe about using the term 'local memory', and he pointed out it's overly restrictive (this technique can apply to private memory, and to device globals). We agreed that the term 'onchip memory' is more accurate.
> 
> I still feel that 'onchip_memory_cache' is a bit ambiguous - it could mean just the cache is in onchip memory and the memory itself is off chip, or it could mean the memory BEING cached is on chip. So I've renamed my class to 'OnchipMemoryWithCache', and renamed the header file to match.
> 
> I feel like this is close enough to 'onchip_memory_cache' that I don't have to rename the tutorial. I think renaming the tutorial is a bit of a pain and has some trickle-down extra work involved that I'd rather avoid. Let me know if you don't like this compromise.

I think that this is a very good compromise

**Comment by mdbtucker** [comment]
@yuguen-intel I've updated the README files (just need to do a HW run to capture the expected output), could you take a look when you have time?

**Comment by mdbtucker** [comment]
> You also probably need to update the MSVS files to have the added include path to the include folder

I've updated the vcxproj file to add the include directory, but since regtests don't currently support testing the MSVC files, I have no way to test this.

I also updated the .json file to suppor the include directory, but since the CI is currently not working, I also have no way to test this change.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md [hunk 8]
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
 
 When the loop contains a loop-carried variable implemented in on-chip memory, the compiler often *cannot* achieve II=1 because the memory access takes more than one clock cycle. If the updated memory location may be needed on the next loop iteration, the next iteration must be delayed to allow time for the update, hence II > 1.
 
-The on-chip memory cache technique breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
+The technique of using the OnchipMemoryWithCache class breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md]:
```suggestion
The technique of using the `OnchipMemoryWithCache` class breaks this dependency by storing recently-accessed values in a cache capable of a 1-cycle read-modify-write operation. The cache is implemented in FPGA registers rather than on-chip memory. By pulling memory accesses preferentially from the register cache, the loop-carried dependency is broken.
```

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md [hunk 9]
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
...
 
-### When is the on-chip memory cache technique applicable?
+### When is the on-chip memory with cache technique applicable?
 
 ***Failure to achieve II=1 because of a loop-carried memory dependency in on-chip memory***:
-The on-chip memory cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
+The on-chip memory with cache technique is applicable if the compiler could not pipeline a loop with II=1 because of an on-chip memory dependency. (If the compiler could not achieve II=1 because of a *global* memory dependency, this technique does not apply as the access latencies are too great.)
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: I think this would benefit from a small graphic showing how the load/store latency of the memory gets on the loop-carried feedback path.

Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: I assume that when IP Authoring is no longer a mischievous glimmer in the eyes of our engineers, we can generalize this from "on-chip memory" to "any *fixed-latency* memory accesses"? Functionally speaking, I realize that these are the same, but in practice this means that kernels that read-modify-write to a fixed-latency MM Host interface could also benefit from this design pattern.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md [hunk 10]
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
...
 
 To check this for a given design, view the "Loops Analysis" section of its optimization report. The report lists the II of all loops and explains why a lower II is not achievable. Check whether the reason given resembles "the compiler failed to schedule this loop with smaller II due to memory dependency". The report will describe the "most critical loop feedback path during scheduling". Check whether this includes on-chip memory load/store operations on the critical path.
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:31]: Is there a way we can put a screenshot in here? I REALLY don't like describing what the report looks like in words when a picture would remove ambiguity. I know pictures are tough to maintain because when the GUI changes the picture needs to be re-screenshot, but that should be easy to do since these code samples are all regtested.

**Comment by yuguen** [comment]
> > You also probably need to update the MSVS files to have the added include path to the include folder
> 
> I've updated the vcxproj file to add the include directory, but since regtests don't currently support testing the MSVC files, I have no way to test this.

Yeah it's quite a pain, but if you look at the qrd one for example, you should have the correct way of doing it. I did test the MSVS files for the qrd on my laptop install of oneapi back then

> 
> I also updated the .json file to suppor the include directory, but since the CI is currently not working, I also have no way to test this change.

~~This will be tested when you'll run the regtest~~ You are right, as long as the CI is down we won't be able to test this

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 9 above)
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: Just a general question: can't we use a very large cache in that case? (e.g. 300 entries)

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json [hunk 11]
```diff
@@ -5,9 +5,18 @@
   "description": "Intel® FPGA tutorial demonstrating the caching of on-chip memory to reduce loop initiation interval",
   "toolchain": ["dpcpp"],
   "os": ["linux", "windows"],
-  "builder": ["ide", "cmake"],
   "targetDevice": ["FPGA"],
+  "builder": ["ide", "cmake"],
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:9]: I'm not sure if it makes a difference, but you moved the builder entry one line down.
It should not make a difference, but to avoid braking a flow somewhere, we might as well let it where it was before

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json [hunk 12]
```diff
@@ -5,9 +5,18 @@
...
   "languages": [{"cpp":{}}],
+  "commonFolder": {
+    "base": "../../..",
+    "include": [
+      "Tutorials/DesignPatterns/onchip_memory_cache",
+      "include/onchip_memory_with_cache.hpp",
+      "include/unrolled_loop.hpp"
```
Comment (yuguen) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:16]: Looking at what is done in qrd, it only includes "include" rather than individual files.
As we can't test this file right now, I'd suggest doing the same as what we know worked before here

Comment (yuguen) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:16]: Looking at the merge sort design, it lists the required include files, so you are all good

File: DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp (hunk 5 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp]: There is only a valid per cache entry.  If all the cache entries are invalid, then a read will pull data from the actual memory.  If the data in the actual memory has not been initialized, then you read garbage.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 9 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: We could, but I'm pretty sure it would not be efficient/performant to create a 300 entry cache with this method (purely register backed).  A cache for off-chip memory would probably need a combination of local memory and registers in the cache.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json (hunk 11 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:9]: I did this to match the sample.json I was using as a reference (pipe_array), which I know passes the CI.

File: DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp (hunk 5 above)
Comment (yuguen) [reply DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp]: Oh right, sorry for some reason I was thinking that the initialization was for the cache values! My bad!

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json (hunk 11 above)
Comment (yuguen) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/sample.json:9]: Understood, then that's fine

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/CMakeLists.txt (hunk 2 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/CMakeLists.txt:47]: When -qactypes is added to the REPORT_LINK_FLAGS, it generates a warning:

/p/psg/swip/w/mtucker/oneapi_samples_mdbtucker.1/DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/build$ make report
Scanning dependencies of target onchip_memory_cache_report.a
[ 50%] Building CXX object src/CMakeFiles/onchip_memory_cache_report.a.dir/onchip_memory_cache.cpp.o
[100%] Linking CXX executable ../onchip_memory_cache_report.a
dpcpp: warning: argument unused during compilation: '-qactypes' [-Wunused-command-line-argument]

That's why I only add it when compiling the fpga target.

**Review by yuguen — APPROVED** [review]
Looks good to me :)

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md [hunk 13]
```diff
@@ -21,36 +21,36 @@ In DPC++ task kernels for FPGA, our objective is to achieve an initiation interv
...
+For user designs, each iteration takes only a few moments to compile the reports. It is important to find the *minimal* value of `CACHE_DEPTH` that results in a maximal performance increase. Unnecessarily large values of `CACHE_DEPTH` consume unnecessary FPGA resources and can reduce f<sub>MAX</sub>. Therefore, at a `CACHE_DEPTH` that results in II=1 and load latency = 1, if further increases to `CACHE_DEPTH` show no improvement, `CACHE_DEPTH` should not be increased any further.
 
-Two versions of the histogram kernel are implemented in the tutorial: one with and one without caching. The report shows II > 1 for the loop in the kernel without caching and II = 1 for the one with caching.
+This tutorial creates multiple kernels sweeping across different cache depths within a single design.  This allows a single compile of the reports to determine the optimal cache depth.
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:53]: this would be a great place for a graph :)

File: DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+    cache_addr_[k_cache_depth-1] = addr;
+    cache_valid_[k_cache_depth-1] = true;
+  }
+
+  T read(addr_t addr) { 
+    T return_val;
+    bool in_cache = false;
+
+    // check the cache, newest value will take precedence
+    #pragma unroll
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:73]: do we not have an `[[intel::unroll]]` attribute?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 15]
```diff
@@ -5,140 +5,97 @@
...
-
-      // Initialize Histogram to zero
-      for (uint32_t b = 0; b < kNumOutputs; ++b) {
-        local_output[b] = 0;
-        local_output_with_cache[b] = 0;
-      }
-
+      // A k_cache_depth of 0 is equivalent to a standard array with no cache
+      fpga_tools::OnchipMemoryWithCache<uint32_t, kNumOutputs, k_cache_depth> 
+        histogram(0);
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:45]: can you call this `histogram_mem` so it's easier to see later in the code that it's a memory? (I'm stupid and need to be beat over the head with information so it sinks in.)
```suggestion
        histogram_mem(0);
```

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 16]
```diff
@@ -5,140 +5,97 @@
...
-
-          // Cache is just a shift register, so shift the shift reg. Pushing
-          // into the back of the shift reg is done above.
-          #pragma unroll
-          for (int i = 0; i < kCacheDepth; i++) {
-            last_sum[i] = last_sum[i + 1];
-            last_sum_index[i] = last_sum_index[i + 1];
-          }
-        }
+      for (uint32_t n = 0; n < kInitNumInputs; ++n) {
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:47]: Why not use `[[intel::iteration_interval 1]]` to force this loop to achieve II=1? Or is the problem that it doesn't achieve II=1 because you can't use `ivdep` safely as has been discussed ad nauseum?

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp [hunk 17]
```diff
@@ -5,140 +5,97 @@
...
-  e = q.submit([&](handler& h) {
+  e = q.submit([&](sycl::handler& h) {
     // Get accessors to the SYCL buffers
-    accessor input(input_buf, h, read_only);
-    accessor output(output_buf, h, write_only, no_init);
+    sycl::accessor input(input_buf, h, sycl::read_only);
+    sycl::accessor output(output_buf, h, sycl::write_only, sycl::no_init);
 
-    h.single_task<Task<use_cache>>([=]() [[intel::kernel_args_restrict]] {
+    h.single_task<HistogramID<k_cache_depth>>(
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:39]: this is a nice clean kernel.

File: DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    }
+  }
+
+  OnchipMemoryWithCache(const OnchipMemoryWithCache&) = delete;
+  OnchipMemoryWithCache& operator=(const OnchipMemoryWithCache&) = delete;
+
+  // explicitly communicate to developers that we don't want to support a
+  // square bracket operator that returns a reference, as it would allow
+  // modification of the memory without updating the cache
+  template <typename I> T& operator[](I addr) = delete;
```
Comment (whitepau) [inline DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:39]: Why not implement a square-bracket operator that simply calls the `write()` function like you did for the `read()`? I think `[]` is a better API than `read()` and `write()`.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 9 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: I'm not sure what such a diagram would look like, and I'm afraid I've spent more time on this already than I had planned, so that will have to wait for a future update (perhaps by you?)

Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:29]: I guess the technique could work, but this abstracted class would not, as it contains the memory itself.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 10 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:31]: I don't think we have screenshots of reports in any other code sample.  I agree with you it's preferable to the user, but I think the burden of keeping the screen shots up to date is excessive.

File: DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp (hunk 14 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:73]: Hmm, I never thought of that, but it's a perfectly reasonable question.  As far as I can tell, the answer is no such attribute exists.  If it does exist, it's very hard to find in our docs, and the loop_unroll code sample does not mention it.

I've reached out to the loops team to ask WHY that is the case, I will let you know the answer (if there is one).

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 13 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md:53]: It would be a pretty boring graph, since the throughput is the same for all the kernels except the one without a cache.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 15 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:45]: It's not just a memory though, it's a memory with a cache.

Also, I don't think it's standard practice to add _mem to anything that's a memory.  Most arrays aren't called, for example, q_matrix_mem or steering_vector_mem.  I don't think this would enhance the clarity of the code.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 16 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:47]: It can't achieve II1 for the case with 0 depth cache (no cache).  So I can't add an II1 here.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp (hunk 17 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/src/onchip_memory_cache.cpp:39]: I've got global memory read and write pointers that don't alias, so I think the kernel_args_restrict is appropriate here.

File: DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp (hunk 18 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/include/onchip_memory_with_cache.hpp:39]: Normally a square bracket operator returns a reference.  I can't do that here, I need a function call to read, and a different function call to write.  I don't know if that's possible with a singler operator, basically I need to do different things depending on if it's being used on the LHS or RHS of an operation.  If it is possible, it's beyond my C++ skillz.

File: DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md (hunk 8 above)
Comment (mdbtucker) [reply DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/onchip_memory_cache/README.md]: Fixed!
